src/main/kotlin/org/jitsi/videobridge/transport/octo/OctoTransport.kt

```diff
@@ -0,0 +1,269 @@
+    }
+
+    private fun sendData(
+        buf: ByteArray,
+        off: Int,
+        len: Int,
+        targets: Set<SocketAddress>,
+        confId: String,
+        mediaType: MediaType,
+        sourceEpId: String? = null
+    ) {
+        val octoPacketLength = len + OCTO_HEADER_LENGTH
+
+        // Not all packets originate from an endpoint (e.g. some come from the bridge)
+        val epId = sourceEpId ?: "ffffffff"
+
+        val (newBuf, newOff) = when {
+            off >= OCTO_HEADER_LENGTH -> {
+                // We can fit the Octo header into the room left at the beginning of the packet
+                Pair(buf, off - OCTO_HEADER_LENGTH)
+            }
+            buf.size >= octoPacketLength -> {
+                // The buffer is big enough to hold the Octo header as well, but we
+                // need to shift it
+                System.arraycopy(buf, off, buf, OCTO_HEADER_LENGTH, len)
+                Pair(buf, 0)
+            }
+            else -> {
+                // The buffer isn't big enough, we need a new one
+                val newBuf = ByteBufferPool.getBuffer(octoPacketLength).apply {
+                    System.arraycopy(buf, off, this, OCTO_HEADER_LENGTH, len)
+                }
+                Pair(newBuf, 0)
+            }
+        }
+        OctoPacket.writeHeaders(
+            newBuf, newOff,
+            true /* source is a relay */,
+            mediaType,
+            0 /* simulcast layers info */,
+            confId,
+            epId
+        )
+        outgoingDataHandler?.sendData(newBuf, newOff, octoPacketLength, targets) ?: stats.noOutgoingHandler()
+    }
+
+    fun getStats(): OrderedJsonObject = OrderedJsonObject().apply {
+        put("relay_id", relayId)
+        putAll(stats.toJson())
+    }
+
+    private fun createPacketInfo(sourceEpId: String, buf: ByteArray, off: Int, len: Int, receivedTime: Instant): OctoPacketInfo {
+        val rtpLen = len - OCTO_HEADER_LENGTH
+        val bufCopy = ByteBufferPool.getBuffer(
+            rtpLen + RtpPacket.BYTES_TO_LEAVE_AT_START_OF_PACKET + Packet.BYTES_TO_LEAVE_AT_END_OF_PACKET
+        ).apply {
+            System.arraycopy(
+                buf, off + OCTO_HEADER_LENGTH,
+                this, RtpPacket.BYTES_TO_LEAVE_AT_START_OF_PACKET,
+                rtpLen
+            )
+        }
+        return OctoPacketInfo(UnparsedPacket(bufCopy, RtpPacket.BYTES_TO_LEAVE_AT_START_OF_PACKET, rtpLen)).apply {
+            this.endpointId = sourceEpId
+            this.receivedTime = receivedTime.toEpochMilli()
+        }
+    }
+
+    private fun createMessageString(buf: ByteArray, off: Int, len: Int): String {
+        return String(
+            buf,
+            off + OCTO_HEADER_LENGTH,
+            len - OCTO_HEADER_LENGTH
+        ).also {
+            logger.cdebug { "Received a message in an Octo data packet: $it" }
+        }
+    }
+
+    private class Stats {
```

Comment:
There's a single `OctoTransport` for the whole bridge, right? Doesn't it need to be thread safe?

Reply:
True...I think the receive-side stats should technically be fine (`UdpTransport` will always call them from the same thread) but maybe I'll tweak all of them for consistency (and since technically it could be fed by something else).